pkg/processor/worker/allocator.go

```diff
@@ -125,3 +136,125 @@ func (fp *fixedPool) Shareable() bool {
 func (fp *fixedPool) GetWorkers() []*Worker {
 	return fp.workers
 }
+
+func (fp *fixedPool) Delete(worker *Worker) error {
+	return ErrNoDelete
+}
+
+// flexible pool
+type flexPool struct {
+	runtimeConfiguration *runtime.Configuration
+	logger               logger.Logger
+	lock                 sync.Mutex
+	freeWorkers          map[*Worker]bool
+	allocatedWorkers     map[*Worker]bool
+}
+
+// NewFlexPoolWorkerAllocator return a new worker pool with flexible size
+func NewFlexPoolWorkerAllocator(parentLogger logger.Logger, runtimeConfiguration *runtime.Configuration, workers []*Worker) (Allocator, error) {
+	configurationCopy := *runtimeConfiguration
+	flexPool := &flexPool{
+		logger:               parentLogger.GetChild("flex_pool_allocator"),
+		runtimeConfiguration: &configurationCopy,
+		freeWorkers:          make(map[*Worker]bool),
+		allocatedWorkers:     make(map[*Worker]bool),
+	}
+
+	for _, worker := range workers {
+		flexPool.freeWorkers[worker] = true
+	}
+
+	return flexPool, nil
+}
+
+func (fa *flexPool) nextIndex() int {
+	return len(fa.freeWorkers) + len(fa.allocatedWorkers)
+}
+
+func (fa *flexPool) Allocate(timeout time.Duration) (*Worker, error) {
+	fa.lock.Lock()
+	defer fa.lock.Unlock()
+
+	if len(fa.freeWorkers) == 0 {
+		worker, err := WorkerFactorySingleton.CreateWorker(fa.logger, fa.nextIndex(), fa.runtimeConfiguration)
+		if err != nil {
+			return nil, err
+		}
+
+		fa.freeWorkers[worker] = true
+	}
+
+	for worker := range fa.freeWorkers {
+		delete(fa.freeWorkers, worker)
+		fa.allocatedWorkers[worker] = true
+		return worker, nil
+	}
+
+	return nil, ErrNoAvailableWorkers
+}
+
+func (fa *flexPool) Release(worker *Worker) {
+	fa.lock.Lock()
+	defer fa.lock.Unlock()
+
+	if _, ok := fa.allocatedWorkers[worker]; !ok {
+		fa.logger.WarnWith("Release of non allocated worker", "index", worker.GetIndex())
+		return
+	}
+
+	delete(fa.allocatedWorkers, worker)
+	fa.freeWorkers[worker] = true
+}
+
+func (fa *flexPool) Shareable() bool {
+	return true
+}
+
+func (fa *flexPool) GetWorkers() []*Worker {
+	fa.lock.Lock()
+	defer fa.lock.Unlock()
+
+	workers := make([]*Worker, 0, len(fa.freeWorkers)+len(fa.allocatedWorkers))
+	for worker := range fa.freeWorkers {
+		workers = append(workers, worker)
+	}
+
+	for worker := range fa.allocatedWorkers {
+		workers = append(workers, worker)
+	}
+
+	return workers
+}
+
+func (fa *flexPool) SetWorkers(workers []*Worker) error {
+	fa.lock.Lock()
+	defer fa.lock.Unlock()
+
+	fa.freeWorkers = make(map[*Worker]bool)
+	fa.allocatedWorkers = make(map[*Worker]bool)
+
+	for _, worker := range workers {
+		fa.freeWorkers[worker] = true
+	}
+
+	return nil
+}
+
+func (fa *flexPool) Delete(worker *Worker) error {
```

Comment:
Please add unit tests for this

Reply:
As in [here](https://github.com/tebeka/nuclio/blob/flex-allocator/pkg/processor/worker/allocator_test.go#L144) or do you have more in mind?

Reply:
I mean in allocator tests below. I still am not sure how the next index thing works - it should also be unit tested in the allocator tests. allocate a few, delete, etc - make sure index is unique and as expected